pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java

```diff
@@ -475,6 +476,10 @@ private static void getSegmentExplainPlanRowData(Operator node, ExplainPlanRows
       }
     }
 
+    if (node instanceof AcquireReleaseColumnsSegmentOperator) {
+      ((AcquireReleaseColumnsSegmentOperator) node).materializeChildOperator();
```

Comment:
I think this will cause exception without first calling `acquire()`

Reply:
Are you saying this because the run() method for some PlanNode can read actual buffer to generate the operator?
While that is true (case in point: https://github.com/apache/pinot/blob/master/pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java#L291), prefetch / acquire will only make whole thing faster by fetching those buffers everything in advance right? Even if we refer to the buffers without prefetch / acquire, it'll still work (albeit slow)? I was able to get correct plan outputs with both the SegmentLocalFSDirectory (where prefetch / acquire are NO-OP), and some other implementations of the SegmentDirectory where these functions are implemented.

I guess the only time this will fail is if some SegmentDirectory implementation necessarily needs prefetch / acquire before being able to access the buffers at all?

Reply:
cc @npawar for more input.
The reason why we need to materialize child operator within the next block call instead of during the planing phase is because we want to acquire first. This PR breaks that use pattern

Reply:
Hi @saurabhd336 , 
Jackie and I just discussed this. 
Ideally we want the contract to be, that prefetch/acquire/release has happened before any physical access related to the query. While just accessing planNodes (which is the case for explain plan, similar to our planning phase) shouldn't really do physical access, we previously discovered that it does do physical access (as you pointed out in your example above). So in a sense, the planning we have is "physical planning", and the explain plan is actually an "explain physical plan". 
In case of the planning part, we overcame this by passing the planNode to the AcquireReleaseOperator and having it run within the getNextBlock, after prefetch/acquire has already happened.
It would be good if we can maintain the same contract for the planNode.run that's happening in explain physical plan i.e. ensure prefetch/acquire is done and a release is called after.
Could we add a new API to all operators, something like "getExplainPlan" and move all this logic into the respective operators? That way we'd avoid all of this if-else happening here (even for EmptyFilterOperator and MatchAllFilterOperator). In the InstanceResponseOperator's getExplainPlan method, we'd prefetch and release (similar to getNextBlock), and in AcquireRelease we'd acquire/release

Reply:
Seems fair that we try and maintain the prefetch -> acquire -> access -> release protocol with this operator. Although I always thought of prefetching as an optimization rather than a mandation so thought it should be okay here to access without prefetching first for the explain plan queries.

From the performance perspective I do have a question though, the `_fetchContext` that the `AcquireReleaseColumnsSegmentOperator` is initialized with, is from https://github.com/apache/pinot/blob/master/pinot- core/src/main/java/org/apache/pinot/core/query/prefetch/FetchPlanner.java#L52. 
while the `PlanNode.run()` may not need all these column-index buffers to run. So do you suggest we change the logic in `FetchPlanner` to be aware of when a query is 'explain plan' and not include all columns?